src/main/java/dev/j3fftw/litexpansion/items/Wrench.java

```diff
@@ -0,0 +1,144 @@
+        ) {
+            e.setCancelled(true);
+            Block block = e.getClickedBlock();
+            SlimefunItem slimefunBlock = BlockStorage.check(block);
+
+            // Check if slimefunBlock is not a machine or a cargo component
+            if (slimefunBlock == null
+                || (!(slimefunBlock instanceof EnergyNetComponent)
+                && !slimefunBlock.getID().startsWith("CARGO_NODE")
+                && !slimefunBlock.getID().equals(SlimefunItems.TRASH_CAN.getItemId()))
+            ) {
+                p.sendMessage(ChatColor.RED + "Hey buddy boy, this can only be electric machines, generators, capacitors, and cargo nodes!");
+                return;
+            }
+
+            // Check if the config states that player needs wrench
+            if (Constants.MACHINE_BREAK_REQUIRES_WRENCH) {
+                double failChance = Math.random();
+                if (failChance <= Constants.WRENCH_FAIL_CHANCE) {
+                    dropBlock(e, p, block, slimefunBlock, true);
+                } else {
+                    dropBlock(e, p, block, slimefunBlock, false);
+                }
+            } else {
+                dropBlock(e, p, block, slimefunBlock, false);
+            }
+            damageItem(p, e.getItem());
+        }
+    }
+
+    public static void dropBlock(Event e, Player p, Block block, SlimefunItem slimefunBlock, boolean failed){
+        ItemStack slimefunBlockDrop = slimefunBlock.getItem();
+        BlockStorage.clearBlockInfo(block);
+        block.setType(Material.AIR);
+
+        // Special cases, capacitors should not turn into machine blocks and cargo nodes do not have input slots
+        if (Arrays.asList(specialBlockIds).contains(slimefunBlock.getID())) {
+
+            // These blocks can not fail
+            block.getLocation().getWorld().dropItemNaturally(block.getLocation(), slimefunBlockDrop);
+
+            if (slimefunBlock.getID().equals(SlimefunItems.CARGO_INPUT_NODE.getItemId()) || slimefunBlock.getID().equals(SlimefunItems.CARGO_OUTPUT_NODE_2.getItemId())) {
+                for (int slot : cargoSlots) {
+                    ItemStack cargoDrop = BlockStorage.getInventory(block).getItemInSlot(slot);
+                    if (cargoDrop != null) {
+                        block.getLocation().getWorld().dropItemNaturally(block.getLocation(), cargoDrop);
+                    }
+                }
+            }
+
+        // All other blocks will be EnergyNetComponents
+        } else {
+            if (failed) {
+                block.getLocation().getWorld().dropItemNaturally(block.getLocation(), Items.MACHINE_BLOCK);
+                p.sendMessage(ChatColor.RED + "Oh no! Your wrench failed!");
+            } else {
+                block.getLocation().getWorld().dropItemNaturally(block.getLocation(), slimefunBlockDrop);
+            }
+
+            if (BlockStorage.hasInventory(block)) {
+                int[] inputSlots = ((InventoryBlock) slimefunBlock).getInputSlots();
```

Comment:
Not all Slimefun Items that have an Inventory are an instance of `InventoryBlock`, in fact only some basic machines use that interface. The interface is only used as a helper utility to provide the generic inventory you sometimes see. But for any more complex interfaces... it is almost never used.

Reply:
Like which? Ive tested a variety of machines that dont have the generic 2 input 2 output layout